system-helper/org.freedesktop.Flatpak.rules.in

```diff
@@ -7,6 +7,12 @@ polkit.addRule(function(action, subject) {
         subject.active == true && subject.local == true &&
         subject.isInGroup("@privileged_group@")) {
             return polkit.Result.YES;
+    } else if ((action.id == "org.freedesktop.Flatpak.app-install" ||
+                action.id == "org.freedesktop.Flatpak.install-bundle") &&
+               subject.active == true && subject.local == true &&
+               action.lookup("parental-controls-app-installation-allowed") === "1" &&
+               action.lookup("parental-controls-is-appropriate") === "1") {
+            return polkit.Result.YES;
```

Comment:
Currently the default for the `allow-app-installation` per-user key in accountsservice is `True` (https://github.com/endlessm/eos-parental-controls/blob/f2e7cbfd0341f17474b1458dc20afc6e67cdf6c8/accounts-service/com.endlessm.ParentalControls.AppFilter.xml#L73). Otherwise, your understanding is correct.

I need to flip that default.

Reply:
https://github.com/endlessm/eos-parental-controls/pull/17

Reply:
If you flip the default, won't it be the case that administrators can install apps system-wide, but not in the user repo?

Reply:
(And if so – is it possible for the unprivileged Flatpak process to ask Polkit what the result of that permission check would be, anyway?)

Reply:
Summary of real-life discussion:

* We want to continue to allow administrator users to install to their user repo by default, but changing the `allow-app-installation` default to `false` would, as this branch stands, break this
  * One reason to care is that GNOME Builder uses the user repo
* We may or may not care whether non-administrator users can install to their user repo; perhaps it is okay to allow it, assuming the OARS rating would allow this app

Possible approaches:

* Rename the `allow-app-installation` key to `allow-system-installation`, and only check it in the system helper
* Check whether the user-repo code could ask Polkit to make the decision
  * With the rule in this PR and `allow-app-installation` defaulting to `false`, this would allow administrators to install to their user repo, but not unprivileged users
  * Unfortunately, the [Polkit docs](https://www.freedesktop.org/software/polkit/docs/latest/polkit-apps.html#polkit-apps-unprivileged) suggest that unprivileged users cannot pass variables to CheckAuthorization, and variables are the mechanism used to pass `allow-app-installation` (and `parental-controls-is-appropriate`) into the Polkit rule

Reply:
Summary of this morning’s Slack discussion:
> ramcq [10:15]
> @pwithnall @wjt discussed with @nedrichards the problem that current defaults (user: yes, system: only if admin) are not compatible with a single boolean which would either increase (if defaulting to true, granting system installs) or reduce (if defaulting to false, breaking user installs) the current privelege levels
> he suggests we go to two booleans, as a parental control system that had a toggle that /didn't/ prevent app installation, when toggled, would be shit
> I suppose the alternative is to invert the sense of the boolean
prevent app installation
> but that doesn't give us a way to selectively grant filtered app installation to a normal user
> which isn't as good for Endless I don't think

> pwithnall [10:20]
> i.e. One boolean for the system repo, one for the user repo?

> wjt [10:25]
> with default False and True respectively?
> What would the UI for this look like? In what contexts would system=True, user=False be sensible?

> ramcq [10:30]
> Nick is pondering that. I would potentially grey out the "Can install system wide?" if "Can install apps?" is N

> ramcq [10:32]
> you could store it as a tristate I guess. or a bitfield. it doesn't /really/ matter. the main thing is we have to store something that allows us to represent the current upstream defaults, and we have to disable user installs if the admin wants to do that

> wjt [10:33]
> Right, I think it should be a tristate: Can't install apps | Can install to user repo, or system if admin | Can install everywhere
> I suppose it's no harder to implement the verification if it's 2 bits rather than a tristate

> ramcq [10:38]
> splitting them felt less magic in some way
> user checks a user setting, polkit checks a system setting

> nedrichards [16:11]
> wireframes updated https://phabricator.endlessm.com/M10

Reply:
>     * Unfortunately, the [Polkit docs](https://www.freedesktop.org/software/polkit/docs/latest/polkit-apps.html#polkit-apps-unprivileged) suggest that unprivileged users cannot pass variables to CheckAuthorization, and variables are the mechanism used to pass `allow-app-installation` (and `parental-controls-is-appropriate`) into the Polkit rule

[The code](https://gitlab.freedesktop.org/pwithnall/polkit/blob/4f795e0d99c64ef8ef309bb76c62e1f08607d9c0/src/polkitbackend/polkitbackendinteractiveauthority.c#L937) backs up the docs, as does a test:
```
$ pkcheck -a org.freedesktop.Flatpak.app-install --details hello test --process $(pidof nautilus)
Error checking for authorization org.freedesktop.Flatpak.app-install: GDBus.Error:org.freedesktop.PolicyKit1.Error.NotAuthorized: Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() and pass details
```

Reply:
Reworked accordingly, but not yet tested.